src/index.d.ts

```diff
@@ -0,0 +1,721 @@
+ * Anything else will be converted to a string and then
+ * turned into a buffer.
+ */
+interface TidyBufferStatic {
+  (document: string | Buffer, options: Generated.OptionDict,
+    callback: TidyCallback): void
+
+  (document: string | Buffer, callback: TidyCallback): void
+}
+
+export class TidyOption {
+  // creation of TidyOption is not exposed
+  private constructor()
+  readonly name: string
+  readonly type: "boolean" | "integer" | "doctype" | "string"
+  readonly readOnly: boolean
+  readonly pickList: string[]
+  readonly id: number
+  readonly category: string
+  toString(): string
+}
+
+/**
+ * TidyDoc is the central object for dealing with the library at a low level
+ *
+ * Such an object will hold a configuration and be able to process one input
+ * file at a time, while multiple such objects can deal with multiple inputs
+ * simultaneously using an independent configuration for each of them.
+ */
+interface TidyDoc extends Generated.TidyDocOption {
+  // Sync calls
+  cleanAndRepairSync(): string
+  parseBufferSync(document: Buffer): string
+  runDiagnosticsSync(): string
+  saveBufferSync(): Buffer
+  // getErrorLog(): string // is not needed: other calls already return log
+
+  // Async calls
+  cleanAndRepair(callback: TidyCallback): void
+  parseBuffer(document: Buffer, callback: TidyCallback): void
+  runDiagnostics(callback: TidyCallback): void
+  saveBuffer(callback: TidyCallback): void
+  tidyBuffer(buf: Buffer, callback: TidyCallback): void
+
+  // batch set/get of options
+  options: Generated.OptionDict
+  // Methods that return TidyOption object
+  getOptionList(): TidyOption[]
+  getOption(option: TidyOption): TidyOption
+  getOption(optionId: number): TidyOption
+  getOption(optionName: string): TidyOption
+  /**
+   * FIXME: does this return the list of related options??
+   */
+  optGetDocLinksList(optionId: number): TidyOption[]
+  optGetDocLinksList(optionName: string): TidyOption[]
+  /* FIXME: what does this do? */
+  optGetCurrPick(key: any): any
+}
+
+/**
+ * Constructor
+ * (Can be used with `new` or normal call)
+ */
+interface TidyDocConstructor {
+  new (): TidyDoc
+  (): TidyDoc
+}
+
+/**
+ * Elements of the compat namespace offer compatibility with other
+ * libtidy bindings for node.
+ */
+interface TidyCompat {
+  libtidy: {
+    tidy: {
+      (text: string, callback: (err: any, html: string) => void): void
+      (text: string, options: any, callback: (err: any, html: string) => void): void
+    }
+  }
+}
+
+// START-GENERATED-NAMESPACE
+// The rest of this file is generated.
+/**
+* Type for libtidy options
+* @generated with /util/gen-typescript-decl.ts
+*/
+declare namespace Generated {
+  interface TidyDocOption {
+    /**
+    * PrettyPrint / indent-spaces (integer)
+    */
+    optGet(key: "indent-spaces" | "indent_spaces" | "IndentSpaces"): number;
+    optSet(key: "indent-spaces" | "indent_spaces" | "IndentSpaces", value: number): void;
```

Comment:
Correct me if I'm wrong, but here is what I imagine: When you read from JSON, the *value* would be of type `any`, but the *key* would be of type `string`, with no restriction as to what kind of string it would be. So you can't use the key from the JSON to configure these, unless you cast the `string` to something more generic. Right? I originally assumed you'd have to cast to the full list of enum values, but probably casting to `any` would be the more practical solution.

Another scenario might be query string parameters, which map string to string. As libtidy can parse strings, that's perfectly fine in terms of types, but would be problematic in a type-checked TypeScript.